pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java

```diff
@@ -0,0 +1,212 @@
+    public int compareTo(@Nonnull ResultPair o) {
+      return Long.compare(o._value, _value);
+    }
+
+    @Override
+    public String toString() {
+      return _key + ": " + _value;
+    }
+  }
+
+  public static void main(String[] args)
+      throws Exception {
+    // Get all resources in cluster.
+    List<String> resourcesInCluster = HELIX_ADMIN.getResourcesInCluster(CLUSTER_NAME);
+    Collections.sort(resourcesInCluster);
+
+    for (String tableName : resourcesInCluster) {
+
+      // Only process on tables with specific keyword.
+      if (tableName.contains("_non_additive_")) {
+
+        // Get the inverted index config.
+        AbstractTableConfig tableConfig = getTableConfig(tableName);
+        IndexingConfig indexingConfig = tableConfig.getIndexingConfig();
+        List<String> invertedIndexColumns = indexingConfig.getInvertedIndexColumns();
+
+        // Handle null inverted index columns.
+        if (invertedIndexColumns == null) {
+          invertedIndexColumns = new ArrayList<>();
+          indexingConfig.setInvertedIndexColumns(invertedIndexColumns);
+        }
+
+        // Remove empty string.
+        int emptyStringIndex;
+        while ((emptyStringIndex = invertedIndexColumns.indexOf("")) != -1) {
+          invertedIndexColumns.remove(emptyStringIndex);
+        }
+
+        // Only add inverted index if there is no existing ones.
+        if (invertedIndexColumns.isEmpty()) {
+
+          // Table must have a schema to add inverted index.
+          Schema tableSchema = getTableSchema(tableName);
+          if (tableSchema != null) {
+
+            // Skip tables without dimensions.
+            List<String> dimensionNames = tableSchema.getDimensionNames();
+            if (dimensionNames.size() == 0) {
+              continue;
+            }
+
+            // Skip tables without proper time column.
+            String timeColumnName = tableSchema.getTimeColumnName();
+            if (timeColumnName == null || !timeColumnName.equals("Date")) {
+              continue;
+            }
+
+            // Only add inverted index to table larger than a threshold.
+            JSONObject queryResponse = sendQuery("SELECT COUNT(*) FROM " + tableName);
+            long numTotalDocs = queryResponse.getLong("totalDocs");
+            if (numTotalDocs > TABLE_SIZE_THRESHOLD) {
+
+              // Get each dimension's cardinality on one day's data.
+              queryResponse = sendQuery("SELECT Max(Date) FROM " + tableName);
+              int maxDate = queryResponse.getJSONArray("aggregationResults").getJSONObject(0).getInt("value");
+              // Query DISTINCTCOUNT on all dimensions in one query might cause timeout, so query them separately.
+              List<ResultPair> resultPairs = new ArrayList<>();
+              for (String dimensionName : dimensionNames) {
+                String query =
+                    "SELECT DISTINCTCOUNT(" + dimensionName + ") FROM " + tableName + " WHERE Date = " + maxDate;
```

Comment:
distinctcount on dimension name can be significant hit (date predict may soften it but not sure...)

Use segment metadata api to get cardinality directly